UCSLibraries/src/main/java/it/cnr/iit/xacml/wrappers/PolicyWrapper.java

```diff
@@ -62,50 +63,70 @@ private PolicyWrapper() {}
+            throw new PolicyException( "Error unmarshalling policy : {0}" + e.getMessage() );
+        }
         policyWrapper.setPolicy( policy );
+        return policyWrapper;
+    }
 
-        return policyWrapper.policyType != null ? policyWrapper : null;
+    public static PolicyWrapper build( PolicyType policyType ) throws PolicyException {
+        PolicyWrapper policyWrapper = new PolicyWrapper();
+        try {
+            String policy = marshalPolicyType( policyType );
+            policyWrapper.setPolicy( policy );
+        } catch( JAXBException e ) {
+            throw new PolicyException( "Error marshalling policy : {0}" + e.getMessage() );
+        }
+        policyWrapper.setPolicyType( policyType );
+        return policyWrapper;
+    }
+
+    public static PolicyWrapper build( PAPInterface pap, TryAccessMessage message ) throws PolicyException {
+        String policy = message.getPolicy();
+        if( policy == null && message.getPolicyId() != null ) {
+            policy = pap.retrievePolicy( message.getPolicyId() );
+        }
+        return PolicyWrapper.build( policy );
     }
 
     @Override
     public List<Attribute> getAttributesForCondition( String conditionName ) {
         Reject.ifBlank( conditionName );
-
-        List<Object> list = policyType.getCombinerParametersOrRuleCombinerParametersOrVariableDefinition();
-
-        for( Object obj : list ) {
-            if( !( obj instanceof RuleType ) ) {
-                continue;
-            }
-
-            RuleType ruleType = (RuleType) obj;
+        for( RuleType ruleType : policyType.getRuleTypeList() ) {
             List<ConditionType> conditions = ruleType.getCondition();
-            if( conditions == null ) {
-                continue;
-            }
-
-            for( ConditionType conditionType : conditions ) {
-                String decisionTime = conditionType.getDecisionTime();
-                if( decisionTime == null ) {
-                    if( conditionName.equals( PolicyTags.CONDITION_PRE ) ) {
-                        return extractAttributes( conditionType );
+            if( conditions != null ) {
+                for( ConditionType conditionType : conditions ) {
+                    List<Attribute> attributeList = getAttributesFromCondition( conditionType, conditionName );
+                    if( !attributeList.isEmpty() ) {
+                        return attributeList;
                     }
-                    return new ArrayList<>();
-                } else if( decisionTime.equals( conditionName ) ) {
-                    return extractAttributes( conditionType );
                 }
             }
         }
-
         log.warning( String.format( MSG_WARN_COND_NOT_FOUND, conditionName ) );
         return new ArrayList<>();
     }
 
+    private List<Attribute> getAttributesFromCondition( ConditionType conditionType, String conditionName ) {
+        String decisionTime = conditionType.getDecisionTime();
```

Comment:
creating a local variable here may be expensive since this method is called from within a loop.
Can you not refer to conditionType.getDecisionTime() directly?

Reply:
ok